x-pack/plugins/stack_connectors/server/connector_types/security/tines/tines.ts

```diff
@@ -0,0 +1,189 @@
+      result.push({ id, name, path, secret, storyId });
+    }
+  });
+}
+
+export class TinesConnector extends SubActionConnector<TinesConfig, TinesSecrets> {
+  private urls: {
+    stories: string;
+    agents: string;
+    getRunWebhookURL: (webhook: TinesWebhookObject) => string;
+  };
+
+  constructor(params: ServiceParams<TinesConfig, TinesSecrets>) {
+    super(params);
+
+    this.urls = {
+      stories: `${this.config.url}${API_PATH}/stories`,
+      agents: `${this.config.url}${API_PATH}/agents`,
+      getRunWebhookURL: (webhook) =>
+        `${this.config.url}${WEBHOOK_PATH}/${webhook.path}/${webhook.secret}`,
+    };
+
+    this.registerSubActions();
+  }
+
+  private registerSubActions() {
+    this.registerSubAction({
+      name: SUB_ACTION.STORIES,
+      method: 'getStories',
+      schema: TinesStoriesActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.WEBHOOKS,
+      method: 'getWebhooks',
+      schema: TinesWebhooksActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.RUN,
+      method: 'runWebhook',
+      schema: TinesRunActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.TEST,
+      method: 'runWebhook',
+      schema: TinesRunActionParamsSchema,
+    });
+  }
+
+  private getAuthHeaders() {
+    return { 'x-user-email': this.secrets.email, 'x-user-token': this.secrets.token };
+  }
+
+  /**
+   * Recursively request pages for given a Tines request, it stops when there's no more pages
+   * or when the `pageLimit` is hit. Returns an array of the responses data.
+   * @param req The parameters that will be used on every pagination request.
+   * @param accumulatePage Function to add each page data into the accumulated result.
+   * @param options The pagination options that can be configured. Optional.
+   * - `pageLimit` the maximum number of pages the function will request. Optional, defaults to `100`.
+   * - `pageSize` the number of results per page. Optional, defaults to `20`.
+   * @returns An array of items of the response schema type
+   */
+  private async paginatedRequest<R extends TinesBaseApiResponse, T>(
+    req: SubActionRequestParams<R>,
+    accumulatePage: (result: T[], pageData: R) => void,
+    { pageLimit = 100, pageSize = 20 }: { pageLimit?: number; pageSize?: number } = {}
+  ) {
+    if (pageLimit <= 0 || pageSize <= 0) return [];
+    let request = { ...req, params: { ...req.params, per_page: pageSize } };
+
+    const result: T[] = [];
+    do {
+      const pageResponse = await this.request<R>(request);
+      const nextPage = pageResponse.data.meta.next_page;
+      accumulatePage(result, pageResponse.data);
+      request = { ...request, url: nextPage ?? '', params: {} }; // all query params are carried by next_page url
+      pageLimit--;
+    } while (pageLimit > 0 && request.url);
+
+    return result;
+  }
+
+  protected getResponseErrorMessage(error: AxiosError): string {
```

Comment:
The framework intends for this method to return a string rather than throw an error. Could we remove the throwing of errors here?